### ISSUE TITLE
Publishing on IzzyOnDroid

### ISSUE THREAD

**Issue by Slion** [Slion/Fulguris#253]
I'm guessing this is the repository entry point:
https://apt.izzysoft.de/fdroid/

@IzzySoft said:

> Further, in a proper repo, updates are taken care for via the associated app (here: one of the [available F-Droid clients](https://android.izzysoft.de/applists/category/named/apps_markets#group_1181)). That's usually also what users head to when looking for apps they might need – so being present there makes it easier to be found. And as you speak about badges: [take a look here](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) :smiley:

> Apart from that: A nice page indeed! I'd definitely point to that as "homepage" from the app description (there are defined fields for several locations; be welcome to use [my repo browser](https://apt.izzysoft.de/fdroid) to check out some examples from different repositories, or check with [F-Droid.org](https://f-droid.org/) for what's shown there).

> Glad to read! I'd just need a way to get hold of the APKs then. The easiest way is finding them attached to their corresponding Github releases. My updater then can automatically take care and grabs new APKs within 24h of your making them available. If that's not an option for you, a "fixed URL" (changing file-names won't do for automated update checks as my updater cannot predict them) returning a proper LAST_MODIFIED header would work as well.

I'm not planning to publish releases on GitHub ATM. However if you really want to invest the time to support XenForo which is the platform I run on slions.net you can take a look at how the update check is done in Fulguris:
https://github.com/Slion/Fulguris/blob/main/app/src/main/java/acr/browser/lightning/browser/activity/BrowserActivity.kt#L3272

Basically XenForo has a REST API that can be used for such purpose, see:
https://xenforo.com/community/pages/api-endpoints/
The resource endpoints are the ones of interest here.

Alternatively maybe the direct download URL works well for you with LAST_MODIFIED:
https://slions.net/resources/fulguris.10/download

**Comment by Slion** [comment]
@IzzySoft Though you should know the download variant I publish on slions.net does include Firebase too. Worst even it is enabled by default 😯 not out of malice, far from it, just because that's the easiest way for me to guarantee some level of stability thanks to automated crash reports. There ought to be an onboarding experience to allow users to opt-out on first start but that's just one of the many things on the roadmap and there is no telling when I'll get around doing it.

**Comment by IzzySoft** [comment]
Thanks for dedicating this issue! Let me pick some of the facts:

* my updater works with the APIs of Github, GitLab and Codeberg, as those are the forges most apps reside. I lack the time to implement all possibilities (which is why e.g. Sourcehut and Sourceforge are missing as well), and definitely cannot implement an API used by just a single app, sorry. So if you do not want to publish releases at Github, that would leave a "magic link" as only remaining approach – i.e. a URL always pointing to the latest APK and giving a proper LAST_MODIFIED header in the end, showing the APK's date. That's the fallback solution I've set up with the updater.
* https://slions.net/resources/fulguris.10/download is a magic link pointing to the latest APK it seems. But unless that APK was really published "right now", the LAST_MODIFIED header is wrong (giving the exact same time as DATE does). See for yourself: `wget --method=HEAD --server-response https://slions.net/resources/fulguris.10/download` gives `Date: Wed, 03 Nov 2021 11:53:40 GMT` and `Last-Modified: Wed, 03 Nov 2021 11:53:40 GMT`. Not sure which part of the web server config would need to be updated for that.
* Firebase being enabled is unfortunate and of course discouraged for FOSS builds, but would not be a show-stopper for my repo (it would raise the `NonFreeDep` and `NonFreeNet` anti-features, though). So to get you started quickly, we could postpone this for later adjustment – or entirely skip it with the fact of a targeted F-Droid.org inclusion (at which point, after an overlap of 7..10 days, I usually unlist apps from my repo again unless explicitly requested otherwise).
* from the other issues (namely #84 and #178): in https://gitlab.com/fdroid/rfp/-/issues/1442 (the F-Droid RFP) the bot somehow got hold of the wrong `applicationId`, namely `acr.browser.lightning` – an app already listed at F-Droid. It hence closed the RFP considering it obsolete. I see no technical reasons so far speaking against a listing at F-Droid, when using the corresponding gradle flavors. Maybe you want to entirely remove the `lightning` flavor, which is commented out anyway? I've meanwhile reopened the RFP including a short explanation.

**TL;DR:** If you could get that LAST_MODIFIED header working properly, I could immediately establish Fulguris in my repo – and we could then continue working on the F-Droid inclusion whenever you feel ready. Urn… I was too fast with that:

```
Offending libs:
---------------
* Crashlytics (/com/crashlytics): Tracking
* Firebase Data Transport (/com/google/android/datatransport): NonFreeNet
* Google Mobile Services (/com/google/android/gms): NonFreeDep
* Firebase (/com/google/firebase): NonFreeNet,NonFreeDep
* Firebase Analytics (/com/google/firebase/analytics): Tracking
```

Do you really need all of that? 2 eyes I can close, but 5 I don't have (limit in my repo is 4 "offenders"). Maybe you could at least kick out Crashlytics? Envisioning a fast inclusion of the FOSS build with F-Droid, I'd then even go with setting updates to manual whenever you ping me, should the LAST_MODIFIED header prove to difficult to fix.

**Comment by IzzySoft** [comment]
> I thought [that](https://github.com/Slion/Fulguris/blob/f08ef79785ebb1934664adfeae372a833361369b/app/build.gradle#L339) should make sure they are only used for the specified flavours

Yes, that applies to your own explicit definitions. But it might well be that one of your dependencies has a dependency which then in turn drags it in nevertheless. We've seen that quite often at F-Droid. Maybe you could check the tree using `gradle :app:dependencies` to turn up the culprit?

As for last-modified, a wild guess: was the current APK uploaded "Sat 23 Apr 2022 05:22:08 PM CEST"? Because that's what the `ETag` points to when interpreted as Unix timestamp, `ETag: "1650727328"`. Looks like what XenForo is trying to accomplish is to fully disable any caching:

```
  Date: Tue, 07 Jun 2022 19:25:15 GMT
  Expires: Thu, 19 Nov 1981 08:52:00 GMT
  Cache-Control: private, no-cache, max-age=0
  Last-Modified: Tue, 07 Jun 2022 19:25:15 GMT
  ETag: "1650727328"
```

You see Date = Last-Modified, Expires in the past, Cache-Control set to no-cache and a max age of 0 – as if they wanted to have only the ETag to be reliably checked against. But if I set up any rules evaluating that instead, it might fail for other apps as they either do not have any ETag, or deal the cards exactly the other way around. One should not use Last-Modified to fool caching IMHO. If it was not modified, there's no reason to re-download – that's just wasting resources. Not blaming you of course, you didn't set it up such :wink:

**Comment by Slion** [comment]
Though technically it looks like they are doing the right thing with [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag). You could cache ETag if any and check if that has changed. If no ETag fallback to date comparison I guess.

**Comment by IzzySoft** [comment]
If there wouldn't be so many different implementations of ETag (I just vaguely remember that from a discussion among F-Droid devs). Check for yourself at Wikipedia: [ETag generation](https://en.wikipedia.org/wiki/HTTP_ETag#ETag_generation):

> Common methods of ETag generation include using a [collision-resistant](https://en.wikipedia.org/wiki/Collision_resistance) [hash function](https://en.wikipedia.org/wiki/Hash_function) of the resource's content, a hash of the last modification timestamp, or even just a [revision number](https://en.wikipedia.org/wiki/Revision_control).

So that it is a "plain Unix timestamp" here is just another "specificum"". I don't know where XenForo took their ideas from, but they are definitely no standard: a last-modified changing without the file itself being modified makes me mistrust every other header they give; who knows what else they fiddled with for purposes unknown?

You will understand that I'll not implement work-arounds for such a buggy behavior of a single provider (as it would certainly break things for others). Though their implementation of ETag might not be violating any specifications (as there seem to be none), interpreting that as timestamp would only work for them – and not even that would be "future-proof" as it then would be ME going against the specs (ETags are to be compared against ETags, not against timestamps or whatever).

As for the dependencies: are you sure you have the correct file provided for download then? If those dependencies are correct, the file retrieved cannot be the `fdroid` flavor but must obviously be the `download` flavor.

**Comment by Slion** [comment]
> As for the dependencies: are you sure you have the correct file provided for download then? If those dependencies are correct, the file retrieved cannot be the fdroid flavor but must obviously be the download flavor.

Right, the download flavour does have firebase… only the fdroid one does not have it. So I guess we are back to base and we need that fdroid build to be published somehow.

As for ETag you would not need to interpret it as a timestamp. Just use it as per ETag specs. However that's currently irrelevant to solve that problem unless we decide to publish the fdroid flavour on slions.net too.

**Comment by IzzySoft** [comment]
The problem is if I shall compare against an ETag, I need to have the "old ETag" to compare against. I do not store ETags anywhere, so I have nothing to compare against. And as pointed out, I will not establish exceptions to exceptions and specific work-arounds for single apps; not only do I lack the capacity for that, but it would unnecessarily bloat my code-base. So it's either a location I can handle, or I cannot handle it, sorry. I could put it into "manual update mode", but then updates are irregular – and you'd need to inform me whenever one is available.

The easiest thing (from my end) would be if you could manage attaching the fdroid flavored APK to the release tags here. There my updater would find it, and we wouldn't need to play ping-pong.

**Comment by Poussinou** [comment]
Is it still a subject, wince we now have this app (f-droid flavor) in the [official repo](https://f-droid.org/packages/net.slions.fulguris.full.fdroid/)?

**Comment by Slion** [comment]
I don't know. Does it make sense?

**Comment by IzzySoft** [comment]
Probably not – but it often depends. [See here](https://gitlab.com/IzzyOnDroid/repo/-/wikis/FAQ) for some background. Feel free to close if you think it's not needed anymore – I won't feel offended :smile:

**Comment by Slion** [comment]
@IzzySoft Can you not just distribute the F-Droid build there too?

**Comment by IzzySoft** [comment]
What sense would that make? The F-Droid build is distributed by F-Droid, and there are plenty of mirrors already. Further, my repo runs on "personal space", so resources are limited – which is why there should be good arguments for keeping "duplicates" in. The "only good reasons" (like faster updates) would be rendered moot (updates would even take longer).

Apart from that, I'd need to implement such "update path" first with my updater. You can already guess why I never did :wink: And why I won't. The builds would be identical, so the mirror system covers them.

**Comment by Slion** [comment]
Fair enough, thanks again for your help @IzzySoft!
